File: cpp/4.cpp
class Solution {
    public:
        double kth(vector<int> & v1,vector<int> & v2, int i, int j, int k) {
            int n = v1.size(), m = v2.size(), inf = 0x7fffffff, mid = k >> 1;
            if(i >= n) return v2[j + k - 1];
            if(j >= m) return v1[i + k - 1];
            if(k == 1) return min(v1[i], v2[j]);
            int v1m = i + mid - 1 >= n ? inf : v1[i + mid - 1];
            int v2m = j + mid- 1 >= m ? inf : v2[j + mid - 1];
            if(v1m < v2m) return kth(v1, v2, i + mid, j, k - mid);
            else return kth(v1, v2, i, j + mid, k - mid);
        }
        double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
            int mn = nums1.size() + nums2.size(), mid = mn >> 1;
            if(mn & 1) return kth(nums1, nums2, 0, 0, mid + 1);
            else return (kth(nums1, nums2, 0, 0, mid) + kth(nums1, nums2, 0, 0, mid + 1)) * 0.5;
        }
};

File: py/4.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        n, m, i, j = len(nums1), len(nums2), 0, 0
        mn = n + m
        k = mn - 1 >> 1
        while k > 0 :
            p = k - 1 >> 1
            if j + p >= m or (i + p < n and nums1[i + p] < nums2[j + p]): i += p + 1
            else : j += p + 1
            k -= p + 1
        s = 0
        if j >= m or (i < n and nums1[i] < nums2[j]) :
            s = nums1[i]
            i += 1
        else :
            s = nums2[j]
            j += 1
        if mn & 1 :
            return s
        else:
            if j >= m or (i < n and nums1[i] < nums2[j]) :
                return (s + nums1[i]) / 2.0
            else :
                return (s + nums2[j]) / 2.0
